### PR TITLE
Make initial HUD and fix pause issues

### PR DESCRIPTION
**Pull request by owalk**
pause now works with the escape key as well as the HUD button for it.

### CODE REVIEW

File: game/HUD.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
+"""This module contains the HUD class."""
+import pygame
+from button import Button
+from score_controller import ScoreController
+from text import Text
+
+
+class HeadsUpDisplay(pygame.sprite.Group):
+    """ScoreController class.
```
Comment (Sean-Nishi) [inline game/HUD.py:9]: Incorrect type

File: game/HUD.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+    LEVEL_HINT_BOX_X_PX = 500
+    LEVEL_HINT_BOX_Y_PX = 600
+    LEVEL_HINT_BOX_HEIGHT = 50
+    LEVEL_HINT_BOX_WIDTH = 400
+
+    PAUSE_BUTTON_X = 700
+    PAUSE_BUTTON_Y = 30
+    PAUSE_BUTTON_WIDTH = 50
+    PAUSE_BUTTON_HEITH = 30
```
Comment (Sean-Nishi) [inline game/HUD.py:61]: Spelling

File: game/HUD.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+    BOX_ALPHA = 100
+
+    def __new__(cls):
+        """Create a singleton object.
+
+        If singleton already exists returns the previous singleton object
```
Comment (Sean-Nishi) [inline game/HUD.py:68]:
```suggestion
If instance already exists returns the previous singleton object
```

File: game/HUD.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+        """
+        if not hasattr(cls, "instance"):
+            cls.instance = super(HeadsUpDisplay, cls).__new__(cls)
+        return cls.instance
+
+    def __init__(
+        self,
+    ):
```
Comment (Sean-Nishi) [inline game/HUD.py:76]:
```suggestion
def __init__(self):
```

File: game/HUD.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    def init_score_info(self):
+        """Initialize the score information area."""
+        score_rect = pygame.Rect(
+            self.SCORE_BOX_X_PX,
+            self.SCORE_BOX_Y_PX,
+            self.SCORE_BOX_WIDTH,
+            self.SCORE_BOX_HEIGHT,
+        )
+        self.add(self.add_box(score_rect, self.SCORE_BOX_HEIGHT, self.SCORE_BOX_WIDTH))
+        self.__current_score_spr = Text(
```
Comment (Sean-Nishi) [inline game/HUD.py:106]:
```suggestion
        self._current_score_spr = Text(
```

File: game/HUD.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+            self.SCORE_BOX_X_PX,
+            self.SCORE_BOX_Y_PX,
+            "Current Score:",
+        )
+        self.add(self.__current_score_spr)
+        self.__boredom_meter_spr = Text(
```
Comment (Sean-Nishi) [inline game/HUD.py:112]: Double underscores used on private vars. Rename symbol

File: game/HUD.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+            self.SCORE_BOX_X_PX,
+            self.SCORE_BOX_Y_PX + 20,
+            "Boredom Meter:",
+        )
+        self.add(self.__boredom_meter_spr)
+        self.__current_score_value_spr = Text(
```
Comment (Sean-Nishi) [inline game/HUD.py:118]: Double underscores used on private vars. Rename symbol

File: game/HUD.py [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+            self.SCORE_BOX_VALUE_X_SCORE,
+            self.SCORE_BOX_Y_PX + 1,
+            str(self._score.current_score),
+        )
+        self.add(self.__current_score_value_spr)
+        self.__boredom_meter_value_spr = Text(
```
Comment (Sean-Nishi) [inline game/HUD.py:124]: Double underscores used on private vars. Rename symbol

File: game/HUD.py [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+            self.SCORE_BOX_VALUE_X_BOREDOM,
+            self.SCORE_BOX_Y_PX + 21,
+            str(self._score.boredom_meter),
+        )
+        self.add(self.__boredom_meter_value_spr)
+
+    def toggle_pause_level(self):
+        """Toggle the pause level value."""
+        self.pause_level = not self.pause_level
```
Comment (Sean-Nishi) [inline game/HUD.py:133]: Private var needs preceding underscore. Also vars are only referenced in this method

File: game/HUD.py [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+
+    def add_box(self, rect, height, width) -> pygame.sprite.Sprite:
+        """Create a background box for HUD elements."""
+        sprite = pygame.sprite.Sprite()
+        sprite.rect = rect
+        sprite.image = pygame.Surface((height, width))
```
Comment (Sean-Nishi) [inline game/HUD.py:139]: New vars need type hints, even if it is obvious like in this spot ☹️

File: game/HUD.py [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+        Create the class as a singleton object
+    __init__(self)
+        Initialize the singleton object's starting instance
+    init_score_info(self)
+        Init score information
+    current_score(self)
+        Get the current score
+    toggle_pause_level(self)
+        Toggle the level pause value
+    add_box(self, rect, height, width)
```
Comment (Sean-Nishi) [inline game/HUD.py:32]: Needs return value

File: game/HUD.py [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+        sprite.image = pygame.Surface((height, width))
+        sprite.image.fill(pygame.Color("grey"))
+        sprite.image.set_alpha(self.BOX_ALPHA)
+        return sprite
+
+    def update(self):
+        """Run the HUD."""
+        super().update()
+        self.__current_score_value_spr.text_str = self._score.current_score
+        self.__boredom_meter_value_spr.text_str = self._score.boredom_meter
```
Comment (Sean-Nishi) [inline game/HUD.py:148]: Only one preceding underscore needed. Rename symbol

File: game/HUD.py [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+        self.add_level_hint()
+
+    def add_level_hint(self):
+        """Display the level hint if needed."""
+        # if text to display, have level hint sprite added
+        if self._show_level_hint:
+            if not self._level_hint_added_flag:
+                self._level_hint_added_flag = True
+
+                level_hint_rect = pygame.Rect(
```
Comment (Sean-Nishi) [inline game/HUD.py:158]: Typehint

File: game/HUD.py [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+
+    Methods
+    -------
+    __new__(cls)
+        Create the class as a singleton object
+    __init__(self)
+        Initialize the singleton object's starting instance
+    init_score_info(self)
+        Init score information
+    current_score(self)
```
Comment (Sean-Nishi) [inline game/HUD.py:28]: Method removed. Docstring update required

File: game/button.py [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+        The x position of the sprite
+    _y: int
+        The y position of the sprite
+    _width: int
+        The width of the button
+    _height: int
+        The height of the button
+    _on_click_function: callable
+        The function to execute on button clicks
+    _onePress: boolean
```
Comment (Sean-Nishi) [inline game/button.py:25]: Name of var has type. Needs snake_case

File: game/button.py [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+        If buttons are pressed once per click
+    _alreadyPressed: boolean
+        If the button is already pressed
```
Comment (Sean-Nishi) [inline game/button.py:28]: Name of var has type. Needs snake_case

File: game/button.py [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+    _buttonText: str
+        Button text to display
```
Comment (Sean-Nishi) [inline game/button.py:30]: Name of var has type. Needs snake_case

File: game/button.py [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+    image: pygame.Surface
+        Image to display
+    rect: pygame.rect
+        Rect to store button position
+    _text_image: pygame.surface
```
Comment (Sean-Nishi) [inline game/button.py:35]: Incorrect type

File: game/button.py [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+        Image holding the Text
+
+    Methods
+    -------
+    __init__(self)
+        Initialize the singleton object's starting instance
+    text_str(self)
+        Get the text string value
+    text_str(self, new_value)
+        Set the text string value
```
Comment (Sean-Nishi) [inline game/button.py:45]: Method names are short enough to type out "string" fully. Want to avoid shorthand

File: game/button.py [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+    def __init__(
+        self,
+        x,
+        y,
+        width,
+        height,
+        buttonText="Button",
+        onclickFunction=None,
+        onePress=False,
+    ):
```
Comment (Sean-Nishi) [inline game/button.py:59]: Parameters require typehints

File: game/button.py [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+        """Initialize a button."""
+        super().__init__()
+        self._x: int = x
+        self._y: int = y
+        self._width: int = width
+        self._height: int = height
+        self._onclickFunction: callable = onclickFunction
```
Comment (Sean-Nishi) [inline game/button.py:66]: Rename symbol to be snake_case

File: game/button.py [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+        self._onePress: bool = onePress
+        self._alreadyPressed: bool = False
+        self._buttonText: str = buttonText
```
Comment (Sean-Nishi) [inline game/button.py:69]: Symbols should be in snake_case. Rename symbol

File: game/button.py [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+
+        self.fillColors = {
+            "normal": "#ffffff",
+            "hover": "#666666",
+            "pressed": "#333333",
+        }
+
+    def update(self):
+        """Update the state of the button as needed."""
+        mousePos: tuple = pygame.mouse.get_pos()
```
Comment (Sean-Nishi) [inline game/button.py:84]: Symbol should be snake_case

File: game/button.py [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+    _buttonText: str
+        Button text to display
+    image: pygame.Surface
+        Image to display
+    rect: pygame.rect
+        Rect to store button position
+    _text_image: pygame.surface
+        Image holding the Text
+
+    Methods
```
Comment (Sean-Nishi) [inline game/button.py:38]: Only method is update(self)

File: game/game.py [hunk 25]
```diff
@@ -37,6 +39,9 @@ def __init__(self):
         )
         pygame.display.set_caption("Reluctant Hero")
         pygame.display.set_icon(self._screen)
+        ScoreController()
```
Comment (Sean-Nishi) [inline game/game.py:42]: Quick question: Are we not assigning a handle to these because they are singleton?

File: game/text.py [hunk 26]
```diff
@@ -0,0 +1,56 @@
...
+
+    def __init__(self, x_pos, y_pos, text):
+        """Construct the first singleton instance."""
+        super().__init__()
+        size = 20
+        self.color = pygame.Color("black")
+        self._text = text
+        self._font = pygame.font.SysFont("Arial", size)
+        self.render_font()
+        self.rect = pygame.Rect(x_pos, y_pos, 50, 50)
```
Comment (Sean-Nishi) [inline game/text.py:35]: Typehints required

File: game/text.py [hunk 27]
```diff
@@ -0,0 +1,56 @@
...
+
+class Text(pygame.sprite.Sprite):
+    """Text class.
+
+    Class for holding in game text.
+
+    Attributes
+    ----------
+    _current_score: int
+        The current score for the game
```
Comment (Sean-Nishi) [inline game/text.py:13]: Attributes do not match what is in init

File: game/text.py [hunk 28]
```diff
@@ -0,0 +1,56 @@
...
+
+    Methods
+    -------
+    __init__(self)
+        Initialize the singleton object's starting instance
+    text_str(self)
```
Comment (Sean-Nishi) [inline game/text.py:19]: Return type required

**Review by Sean-Nishi — CHANGES_REQUESTED** [review]
Just the comments I added

File: game/game.py (hunk 25 above)
Comment (owalk) [reply game/game.py:42]: Yeah, The purpose of these lines of code is to run the initializer before any of the other classes need to use these singletons. There is no need to access them in the main method, so adding a way to access them here would add clutter to the main method.

File: game/text.py (hunk 28 above)
Comment (owalk) [reply game/text.py:19]: I hadn't been adding the return type to the class doc strings since it's in the method doc strings.
I'll add this in

Comment (owalk) [reply game/text.py:19]: but only when it is not None

File: game/HUD.py (hunk 5 above)
Comment (owalk) [reply game/HUD.py:106]: aren't we mangling private vars and using a single underscore on protected vars?

Comment (owalk) [reply game/HUD.py:106]: I think a bunch of our single underscores should be double and we might want a new issue for this. it's marginal benefits and doesn't matter too much in the grand scheme of things

File: game/HUD.py (hunk 9 above)
Comment (owalk) [reply game/HUD.py:133]: this is the attribute defined on lines 213 and 218. It's used in main and level.py

File: game/HUD.py (hunk 10 above)
Comment (owalk) [reply game/HUD.py:139]: wait the only new var here is sprite :O I think we just need the one type hint on line 137 for these

File: game/HUD.py (hunk 12 above)
Comment (owalk) [reply game/HUD.py:148]: I'm gonna wait for you to respond to my comments on should we mangle private vars before changing these

File: game/HUD.py (hunk 3 above)
Comment (owalk) [reply game/HUD.py:68]: spacing change suggestion not added, wording suggestion added

File: game/text.py [hunk 29]
```diff
@@ -0,0 +1,62 @@
...
+
+    Attributes
+    ----------
+    _text: str
+        The text to display
+    _font: pygame.font.Font
+        The font object to build
+    rect: pygame.Rect
+        The sprite location rect
+    image: pygane.Surface
```
Comment (Sean-Nishi) [inline game/text.py:18]: Spelling

**Comment by owalk** [comment]
@Sean-Nishi before I fix the one new comment it would be nice to have the other comments addressed and closed out

File: game/text.py (hunk 29 above)
Comment (owalk) [reply game/text.py]: okay i went ahead and just fixed this, but the previous comments have not been resolved yet

File: game/button.py [hunk 30]
```diff
@@ -0,0 +1,114 @@
...
+        The width of the button
+    _height: int
+        The height of the button
+    _on_click_function: callable
+        The function to execute on button clicks
+    _one_press: boolean
+        If buttons are pressed once per click
+    _already_pressed: boolean
+        If the button is already pressed
+    _button_text: str
```
Comment (Sean-Nishi) [inline game/button.py:29]:
```suggestion
    _button_text_string: str
```

**Review by Sean-Nishi — APPROVED** [review]

**Pull request merged**